engine/src/main/java/org/pentaho/di/job/entries/pgpencryptfiles/GPG.java

```diff
@@ -0,0 +1,607 @@
+   * @param plainText
+   *          input string to encrypt
+   * @param keyID
+   *          key ID of the key in GnuPG's key database to encrypt with
+   * @return encrypted string
+   * @throws KettleException
+   */
+  public String encrypt( String plainText, String keyID ) throws KettleException {
+    return execGnuPG( "-r \"" + keyID + "\" --encrypt ", plainText, false );
+
+  }
+
+  /**
+   * Signs and encrypts a string
+   *
+   * @param plainText
+   *          input string to encrypt
+   * @param userID
+   *          key ID of the key in GnuPG's key database to encrypt with
+   * @param passPhrase
+   *          passphrase for the personal private key to sign with
+   * @return encrypted string
+   * @throws KettleException
+   */
+  public String signAndEncrypt( String plainText, String userID, String passPhrase ) throws KettleException {
+    try {
+      createTempFile( plainText );
+
+      return execGnuPG(
+        "-r \"" + userID + "\" --passphrase-fd 0 -se \"" + getTempFileName() + "\"", passPhrase, false );
+    } finally {
+
+      deleteTempFile();
+    }
+
+  }
+
+  /**
+   * Sign
+   *
+   * @param stringToSign
+   *          input string to sign
+   * @param passPhrase
+   *          passphrase for the personal private key to sign with
+   * @throws KettleException
+   */
+  public String sign( String stringToSign, String passPhrase ) throws KettleException {
+    String retval;
+    try {
+
+      createTempFile( stringToSign );
+
+      retval = execGnuPG( "--passphrase-fd 0 --sign \"" + getTempFileName() + "\"", passPhrase, false );
+
+    } finally {
+      deleteTempFile();
+    }
+    return retval;
+  }
+
+  /**
+   * Decrypt a string
+   *
+   * @param cryptedText
+   *          input string to decrypt
+   * @param passPhrase
+   *          passphrase for the personal private key to sign with
+   * @return plain text
+   * @throws KettleException
+   */
+  public String decrypt( String cryptedText, String passPhrase ) throws KettleException {
+    try {
+      createTempFile( cryptedText );
+
+      return execGnuPG( "--passphrase-fd 0 --decrypt \"" + getTempFileName() + "\"", passPhrase, false );
+
+    } finally {
+      deleteTempFile();
+    }
+  }
+
+  /**
+   * Create a unique temporary file when needed by one of the main methods. The file handle is store in tmpFile object
+   * var.
+   *
+   * @param content
+   *          data to write into the file
+   * @throws KettleException
+   */
+  private void createTempFile( String content ) throws KettleException {
+    this.tmpFile = null;
+    FileWriter fw;
+
+    try {
+      this.tmpFile = File.createTempFile( "GnuPG", null );
+      if ( log.isDebug() ) {
+        log.logDebug( BaseMessages.getString( PKG, "GPG.TempFileCreated", getTempFileName() ) );
+      }
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG, "GPG.ErrorCreatingTempFile" ), e );
+    }
+
+    try {
+      fw = new FileWriter( this.tmpFile );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fw*. The resource is referenced at lines 569, 570. The resource is closed at line 571. There are other execution paths that do not contain closure statements, for example, when OutputStreamWriter.flush throws an exception. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).